### PR TITLE
Explicit implicits

### PR DESCRIPTION
**Pull request by mlachkar**
Only add implicit parameters. 
The print for these parameters is not exhaustive (only add static symbols, and local symbols). We should make work for non-static values.

This rule still needs testing.

### CODE REVIEW

File: scalafix/rules/src/main/scala/migrate/ExplicitImplicitsRule.scala [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+        Patch.empty
+      } else Patch.empty
+    }.toList.asPatch
+  }
+
+  def getImplicitParams(originalTree: Tree)(implicit unit: g.CompilationUnit): Option[List[String]] =
+    for {
+      term       <- SyntheticHelper.getTermName(originalTree)
+      context    <- CompilerService.getContext(term, g)
+      globalTree <- getTreeFromContext(context)
```
Comment (adpi2) [inline scalafix/rules/src/main/scala/migrate/ExplicitImplicitsRule.scala:55]: The same sequence of calls is made in the `MigrationRule` class. I suggest extracting that to a new method in `CompilerService`:

```scala
def getTreeFromContext(term: Term.Name)(implicit g: ScalafixGlobal): Option[g.Tree] =
  getContext.flatMap(getTreeFromContext)
```

**Review by adpi2 — COMMENTED** [review]
This looks very good!

I don't know why but `Incompat1` is not working on my machine. I got:
```
2020.10.29 10:14:40 [INFO] migrate.Main.previewMigration:29:16 - Migrating /home/piquerez/scalacenter/scala-migrat3/input/src/main/scala/Incompat1.scala
2020.10.29 10:14:40 [INFO] migrate.Main.buildMigrationFiles:92:41 - Found 0 patch candidate(s) in 1 file(s) after 176 milliseconds
[info] - Incompat1.scala *** FAILED ***
[info]   scalafix.interfaces.ScalafixException: Cannot apply patch on file /home/piquerez/scalacenter/scala-migrat3/input/src/main/scala/Incompat1.scala
[info]   at migrate.internal.FileMigrationState.$anonfun$previewAllPatches$2(FileMigrationState.scala:25)
[info]   at migrate.utils.ScalaExtensions$OptionExtension$.toTry$extension(ScalaExtensions.scala:35)
[info]   at migrate.internal.FileMigrationState.previewAllPatches(FileMigrationState.scala:25)
[info]   at migrate.internal.FileMigrationState.previewAllPatches$(FileMigrationState.scala:18)
[info]   at migrate.internal.FileMigrationState$Initial.previewAllPatches(FileMigrationState.scala:38)
[info]   at migrate.Main$.$anonfun$compileInScala3$1(Main.scala:62)
[info]   at scala.collection.immutable.List.map(List.scala:246)
[info]   at scala.collection.immutable.List.map(List.scala:79)
[info]   at migrate.Main$.compileInScala3(Main.scala:62)
[info]   at migrate.Main$.$anonfun$previewMigration$3(Main.scala:39)
```

**Comment by adpi2** [comment]
After a `reload; clean; migrate / test` everything is working fine!

File: scalafix/rules/src/main/scala/migrate/ExplicitImplicitsRule.scala [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    doc.synthetics.collect { case syn @ ApplyTree(function, arguments) =>
+      // case for implicit params
+      if (syn.toString.startsWith("*")) {
+        (for {
+          originalTree <- SyntheticHelper.getOriginalTree(syn)
+          args         <- getImplicitParams(originalTree)
+        } yield Patch.addRight(originalTree, "(" + args.mkString(", ") + ")")).getOrElse(Patch.empty)
+      } else if (syn.toString().contains("(*)")) {
+        Patch.empty
+      } else Patch.empty
```
[the remark this answers is not in the document]
Comment (mlachkar) [reply scalafix/rules/src/main/scala/migrate/ExplicitImplicitsRule.scala:47]: yes. synthetics for conversions are always SYNTHETICS(OriginalTree). Not done yet

File: scalafix/rules/src/main/scala/migrate/ExplicitImplicitsRule.scala [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    }.toList.asPatch
```
[the remark this answers is not in the document]
Comment (mlachkar) [reply scalafix/rules/src/main/scala/migrate/ExplicitImplicitsRule.scala:48]: ok

File: scalafix/rules/src/main/scala/migrate/ExplicitImplicitsRule.scala [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+      if (syn.toString.startsWith("*")) {
+        for {
+          originalTree <- SyntheticHelper.getOriginalTree(syn)
+          args         <- getImplicitParams(originalTree)
+        } yield Patch.addRight(originalTree, "(" + args.mkString(", ") + ")")
+      }
+      // case for implicit conversions
+      else if (syn.toString().contains("(*)")) {
+        None
+      } else None
```
Comment (adpi2) [inline scalafix/rules/src/main/scala/migrate/ExplicitImplicitsRule.scala:58]: There is no default `case _ => None` and so there is a risk of failure?

**Review by adpi2 — CHANGES_REQUESTED** [review]

**Review by adpi2 — APPROVED** [review]

**Pull request merged**